base/shell/explorer/syspager.cpp

```diff
@@ -178,6 +178,90 @@ class CNotifyToolbar :
     void Initialize(HWND hWndParent, CBalloonQueue * queue);
 };
 
+
+class CHardErrorThread
+{
+    DWORD m_ThreadId;
+    LONG m_bThreadRunning;
+    DWORD m_Status;
+    DWORD m_dwResponseType;
+    CStringW m_Title;
+    CStringW m_Text;
+public:
+
+    CHardErrorThread():
+        m_bThreadRunning(FALSE),
+        m_Status(NULL),
+        m_dwResponseType(NULL),
+        m_ThreadId(NULL)
+    {
+    }
+
+    ~CHardErrorThread()
+    {
+        if (m_bThreadRunning)
+        {
+            /* Try to unstuck Show */
+            PostThreadMessage(m_ThreadId, WM_QUIT, 0, 0);   
+
+            HANDLE hThread = OpenThread(THREAD_ALL_ACCESS, FALSE, m_ThreadId);
+            if (hThread)
+            {
+                WaitForSingleObject(hThread, INFINITE);
+                CloseHandle(hThread);
+            }
+        }
+    }
+
+    VOID ThreadProc()
+    {
+        HRESULT hr;
+        CComPtr<IUserNotification> pnotification;
+
+        OleInitialize(NULL);
+
+        hr = CoCreateInstance(CLSID_UserNotification,
+                              NULL,
+                              CLSCTX_INPROC_SERVER,
+                              IID_PPV_ARG(IUserNotification, &pnotification));
+        if (SUCCEEDED(hr))
```

Comment:
This is better, because it always resets `m_bThreadRunning`, but you lose the `FAILED_UNEXPECTEDLY`